### ISSUE TITLE
`Reset` on overridden component changes the source prototype and is not undoable.

### ISSUE THREAD

**Issue by Selmar** [OurMachinery/themachinery-public#985]
- **Version:** Git Master Branch
- **OS:** Windows

`Reset` on overridden component changes the source prototype and is not undoable.

Repro:
- Open new project
- Open sky entity prototype
- Click on the sun-light entity
- In properties view -> render component, right-click `Filename`, press `Override tree`
- Right-click `Filename` again, press `Reset`
- Observe light.entity was modified and ctrl-z does not fix it

**Comment by niklas-ourmachinery** [comment]
I think what happens here is that after `Override tree`, your setup looks like this:

~~~
sun-light --------------> light
  ||                       ||
render' ----------------> render
  ||                       ||
creation-graph' --->  creation-graph -------> editor-icon
~~~

However, when you then choose `Reset`, you are not resetting `creation-graph'` in `sun-light` to `creation-graph`, you are resetting `creation-graph` in `light` to `editor-icon`.

The reason that this edit can't be undone is that The Machinery's undo-model is document/asset based. I.e. when you are editing a specific asset you will only see undo events related to that asset. Since you are editing `sun-light`, you can't undo the action that was made to `light` (resetting `creation-graph`). If you switch and open the `light` entity, you are able to undo the action.

Admittedly, this is all pretty confusing, and I think we should try to improve this. Some ideas:

* It seems like what `Reset` does here is a bug. I think `Reset` should reset `creation-graph'` to `creation-graph`.
* I don't think it should be possible from the properties tab to edit other assets than the root asset that is selected, since that leads to confusing Undo behavior as in this example. You should be forced to override to a local instance in order to edit it.

I'll assign this to @karl-zylinski-ourmachinery , since he's the one that has been doing the most recent work on this.

**Comment by Selmar** [comment]
Indeed, I was assuming it was not possible to modify the source asset through one of its derivatives. With the exception of propagate, of course.

Your explanation makes me understand some of the `Undo` behaviors a little better. They felt unreliable to me, confusing me every now and then. I suppose it's because I am not always aware which scope I am editing something in, but perhaps there are more issues I am not aware of.

It also makes me wonder: it is not possible to undo a `Propagate` action from the place where the action was done. I tested this, and it seems that modifcations done on a field are all lost once `Propagate` is used. That is, Undo does not affect the value anymore. However, it seems the `Undo` actions for the previous value changes are still in the undo stack (I had to `Undo` several times before it removed the override again). I could not find a way to undo the `Propagate` action, perhaps this is by design?

**Comment by niklas-ourmachinery** [comment]
Yes... so **Propagate** in this case is an action that touches more than one asset... and to Undo those actions you have to go to the Asset Browser.

It's pretty messy...

*Personally*, I've always favored a global undo queue -- where every action made goes into a single undo queue and can be undone from wherever.

However, this has the drawback that you can't keep separate undo queues for separate assets. I.e. if you make bunch of changes to asset A, then go over to asset B and make changes there, you can't then go back to A and undo those changes... you have to undo all the changes you made to B first. And it seems like a lot of people depend on this workflow and really want it.

It might seem to make sense to have undo actions associated with "tabs" instead of assets... i.e. any change made in the properties tab can be undone in the properties tab... but that can be pretty problematic too. For example, when editing an entity, you move around between the Tree/Scene/Properties tabs a lot, and it would be a PITA to remember which tab you have to go to in order to undo something.

To clean up the current situation, maybe we could do something like:

* In an "asset editing tab" (any tab associated with an asset) you can undo any change that touched that asset. (That would handle the Propagate case.)
* In a "non-asset-editing tab" (like the Asset Browser) -- you can undo any change. I.e. while one of these tabs has focus it acts as if you had a global undo queue.
* An "asset editing tab" cannot modify other assets unless the current asset is modified too (i.e. Propagate is ok because it modifies both the current asset and the foreign asset). This makes sure that any change made in a tab can be undone in that tab.

Not sure if this behavior is "intuitive" enough and "simple" enough to work.

**Comment by karl-zylinski-ourmachinery** [comment]
> Personally, I've always favored a global undo queue -- where every action made goes into a single undo queue and can be undone from wherever.
> 
> However, this has the drawback that you can't keep separate undo queues for separate assets. I.e. if you make bunch of changes to asset A, then go over to asset B and make changes there, you can't then go back to A and undo those changes... you have to undo all the changes you made to B first. And it seems like a lot of people depend on this workflow and really want it.

Indeed, I think the moment I have for example a scene and also a material open and work a bit in the scene and then a bit in the material and then a bit in the scene and so on, and then realize I want to undo everything you did to the material last 20 mins, then I would go _nuts_ if it starts undoing in my scene.

> * In an "asset editing tab" (any tab associated with an asset) you can undo any change that touched that asset. (That would handle the Propagate case.)
> 
> * In a "non-asset-editing tab" (like the Asset Browser) -- you can undo any change. I.e. while one of these tabs has focus it acts as if you had a global undo queue.
> 
> * An "asset editing tab" cannot modify other assets unless the current asset is modified too (i.e. Propagate is ok because it modifies both the current asset and the foreign asset). This makes sure that any change made in a tab can be undone in that tab.

We have been discussing changes to the workflows around workspaces a lot lately and I think this ties into that discussion.

The idea that me and @tobias-ourmachinery has been pushing for is that each workspace is tied to a specific asset type, and that each workspace can have several of those assets open at the same time. If you switch between an asset within a workspace, then tabs such as scene tab, properties tab etc all stay within the context of that asset. If it works like this, then each workspace-asset-document could have its own undo stack. Anything that happens in the properties panel or scene view for that asset stays on a undo stack that is unique to that workspace-asset.

Regarding the asset browser, which resides in the Global Tabwell, I don't think it should have a global undo. It should undo anything that has been done within the global tabwell / asset browser. I.e. the global tabwell has 'global tabwell undo stack' that applies to anything that lives in there, but that stack will not get undo items from other tabs outside of it.

This brings us to the weird downside of this approach: What property panel does the asset browser, that lives in the global tabwell, use? The idea we have is that it borrows the asset browser from the current asset-workspace and it is clearly marked as borrowed by some border or something. Anything that happens in the property panel during this borrow ends up in the 'global tabwell undo stack' and can thus be undone from either the asset browser or the property panel (while it is in the borrowed mode).

**Comment by Selmar** [comment]
I do not feel like I have worked with the engine enough to give useful input towards next steps, but I can tell you my perspective from a practical point of view:
- I like the document-based undo stack. A node graph tab should have its own undo stack, like editing a file in Visual Studio. This is what I expected. I think so far I haven't been surprised about *what* was being undone.
- I think actions done through a given view should generally be undoable via that view as well (exceptions being things that aren't undoable in the first place).
- Actions on an instance should not modify the prototype unless the action is clearly modifying said prototype (e.g. `Propagate`).
- We can chalk this up to lack of experience for now, but there is something about selection being part of the undo stack that "annoys" me: there are actions in there that are seemingly useless when undone.

**Comment by Selmar** [comment]
> it can be very infuriating when you accidentally misclick and loose your selection

Yes, I was considering this as well and I thought perhaps a separate undo stack for selections could make sense. But perhaps not. I suspect my main gripe with it is that sometimes I can undo and redo without seeing anything change visually (not even the selection). Perhaps something else is going on there.

> I think the main surprise I usually have is not that what was being undone, but what was *not* being undone.

As I was here!
